File: src/neetcode/Java/_6ZigzagConversion.java
package neetcode.Java;

import static java.lang.Math.max;

public class _6ZigzagConversion {

    public static String convert(String s, int numRows) {
        int n = s.length();
        StringBuffer[] arr = new StringBuffer[numRows];
        for (int i = 0; i < numRows; i++) arr[i] = new StringBuffer();
        int i = 0;
        while (i < n) {
            for (int ind = 0; ind < numRows && i < n; ind++) {
                arr[ind].append(s.charAt(i++));
            }
            for (int ind = numRows - 2; ind > 0 && i < n; ind--) {
                arr[ind].append(s.charAt(i++));
            }
        }
        StringBuffer ans = new StringBuffer();
        for (StringBuffer e : arr) {
            ans.append(e);
        }
        return ans.toString();
    }

    public String convert2(String s, int numRows) {
        if (numRows == 1 || s.length() == 1) return s;

        String rs = "";
        for (int i = 0; i < numRows; i++) {
            int pos = i;
            int pro;
            if (i == 0 || i == numRows - 1) {
                pro = 0;
            } else {
                pro = i;
            }
            while (pos < s.length()) {
                if (pro == 0) {
                    rs = rs + s.charAt(pos);
                    pos = pos + (numRows - 1) * 2;
                } else {
                    rs = rs + s.charAt(pos);
                    pos = pos + (numRows - 1) * 2 - pro * 2;
                    if (pos < s.length()) {
                        rs = rs + s.charAt(pos);
                        pos = pos + pro * 2;
                    }
                }
            }
        }
        return rs;
    }

    public static String convert1(String s, int numRows) {
        System.out.println(1 % 3 + " " + 5 % 3);
        if (s.length() <= 2) return s;
        if (numRows <= 1) return s;
        int len = s.length();
        String ss = "";
        int jump = max(numRows + numRows - 2, 1);
        for (int i = 0; i < len; i = i + jump) {
            ss += s.charAt(i);
        }
        int j = 1;
        while (j < (numRows - 1)) {
            for (int i = j; ; ) {
                if (i < len)
                    ss = ss + s.charAt(i);
                else break;
                i += jump - j * 2;//long jump
                if (i < len)
                    ss = ss + s.charAt(i);//short jump
                else break;
                i += j * 2;
            }
            j++;
        }
        for (int i = numRows - 1; i < len; i = i + jump) {
            ss += s.charAt(i);
        }
        return ss;
    }

    public static void main(String[] args) {
        String s = convert("ABCDEFGHIJKLMNOPQRSTWXYZ", 5);
        System.out.println(s);
    }
}

File: src/neetcode/kotlin/_6ZigzagConversion.kt
package neetcode.kotlin

import kotlin.math.max

class _6ZigzagConversion {
    fun convert2(s: String, numRows: Int): String {
        if (numRows == 1 || s.length == 1) return s

        var rs = ""
        for (i in 0 until numRows) {
            var pos = i
            var pro = if (i == 0 || i == numRows - 1) {
                0
            } else {
                i
            }
            while (pos < s.length) {
                if (pro == 0) {
                    rs = rs + s[pos]
                    pos = pos + (numRows - 1) * 2
                } else {
                    rs = rs + s[pos]
                    pos = pos + (numRows - 1) * 2 - pro * 2
                    if (pos < s.length) {
                        rs = rs + s[pos]
                        pos = pos + pro * 2
                    }
                }
            }
        }
        return rs
    }

    companion object {
        fun convert(s: String, numRows: Int): String {
            val n = s.length
            val arr = arrayOfNulls<StringBuffer>(numRows)
            for (i in 0 until numRows) arr[i] = StringBuffer()
            var i = 0
            while (i < n) {
                run {
                    var ind = 0
                    while (ind < numRows && i < n) {
                        arr[ind]!!.append(s[i++])
                        ind++
                    }
                }
                var ind = numRows - 2
                while (ind > 0 && i < n) {
                    arr[ind]!!.append(s[i++])
                    ind--
                }
            }
            val ans = StringBuffer()
            for (e in arr) {
                ans.append(e)
            }
            return ans.toString()
        }

        fun convert1(s: String, numRows: Int): String {
            println((1 % 3).toString() + " " + 5 % 3)
            if (s.length <= 2) return s
            if (numRows <= 1) return s
            val len = s.length
            var ss = ""
            val jump = max((numRows + numRows - 2).toDouble(), 1.0).toInt()
            run {
                var i = 0
                while (i < len) {
                    ss += s[i]
                    i = i + jump
                }
            }
            var j = 1
            while (j < (numRows - 1)) {
                var i = j
                while (true) {
                    if (i < len) ss = ss + s[i]
                    else break
                    i += jump - j * 2 //long jump
                    if (i < len) ss = ss + s[i] //short jump
                    else break
                    i += j * 2
                }
                j++
            }
            var i = numRows - 1
            while (i < len) {
                ss += s[i]
                i = i + jump
            }
            return ss
        }

        @JvmStatic
        fun main(args: Array<String>) {
            val s = convert("ABCDEFGHIJKLMNOPQRSTWXYZ", 5)
            println(s)
        }
    }
}
